safetensors/src/tensor.rs

```diff
@@ -48,34 +49,126 @@ impl std::fmt::Display for SafeTensorError {
+///        (&self.data).into()
+///    }
+///    fn data_len(&self) -> usize{
+///        self.data.len()
+///    }
+/// }
+/// ```
+///
+/// For a borrowed tensor:
+///
+/// ```rust
+/// use safetensors::tensor::{View, Dtype};
+/// use std::borrow::Cow;
+/// struct Tensor<'data>{ dtype: MyDtype, shape: Vec<usize>, data: &'data[u8]}
+///
+/// # type MyDtype = Dtype;
+/// impl<'data> View for Tensor<'data>{
+///    fn dtype(&self) -> Dtype{
+///        self.dtype.into()
+///    }
+///    fn shape(&self) -> &[usize]{
+///         &self.shape
+///    }
+///    fn data(&self) -> Cow<[u8]>{
+///        self.data.into()
+///    }
+///    fn data_len(&self) -> usize{
+///        self.data.len()
+///    }
+/// }
+/// ```
+///
+/// Now if you have some unknown buffer that could be on GPU for instance,
+/// you can implement the trait to return an owned local buffer containing the data
+/// on CPU (needed to write on disk)
+/// ```rust
+/// use safetensors::tensor::{View, Dtype};
+/// use std::borrow::Cow;
+///
+/// # type MyDtype = Dtype;
+/// # type OpaqueGpu = Vec<u8>;
+/// struct Tensor{ dtype: MyDtype, shape: Vec<usize>, data: OpaqueGpu }
+///
+/// impl View for Tensor{
+///    fn dtype(&self) -> Dtype{
+///        self.dtype.into()
+///    }
+///    fn shape(&self) -> &[usize]{
+///         &self.shape
+///    }
+///    fn data(&self) -> Cow<[u8]>{
+///        // This copies data from GPU to CPU.
+///        let data: Vec<u8> = self.data.to_vec();
+///        data.into()
+///    }
+///    fn data_len(&self) -> usize{
+///        let n: usize = self.shape.iter().product();
+///        let bytes_per_element = self.dtype.size();
+///        n * bytes_per_element
+///    }
+/// }
+/// ```
+pub trait View {
+    /// The `Dtype` of the tensor
+    fn dtype(&self) -> Dtype;
+    /// The shape of the tensor
+    fn shape(&self) -> &[usize];
+    /// The data of the tensor
+    fn data(&self) -> Cow<[u8]>;
+    /// The length of the data, in bytes.
+    /// This is necessary as this might be faster to get than `data().len()`
+    /// for instance for tensors residing in GPU.
+    fn data_len(&self) -> usize;
+}
+
+fn prepare<S: AsRef<str> + Ord + std::fmt::Display, V: View, I: IntoIterator<Item = (S, V)>>(
+    data: I,
+    data_info: &Option<HashMap<String, String>>,
     // ) -> Result<(Metadata, Vec<&'hash TensorView<'data>>, usize), SafeTensorError> {
-) -> Result<PreparedData<'hash, 'data>, SafeTensorError> {
+) -> Result<(PreparedData, Vec<V>), SafeTensorError> {
     // Make sure we're sorting by descending dtype alignment
     // Then by name
-    let mut data: Vec<_> = data.iter().collect();
+    let mut data: Vec<_> = data.into_iter().collect();
     data.sort_by(|(lname, left), (rname, right)| {
-        right.dtype.cmp(&left.dtype).then(lname.cmp(rname))
+        right.dtype().cmp(&left.dtype()).then(lname.cmp(rname))
     });
 
-    let mut tensors: Vec<&TensorView> = vec![];
+    let mut tensors: Vec<V> = vec![];
```

Comment:
You can do `Vec::with_capacity` like below

Reply:
Thanks.